dash-pipeline/bmv2/dash_tunnel.p4

```diff
@@ -0,0 +1,280 @@
+    hdr. ## overlay_id ## _ethernet.dst_addr = overlay_dmac; \
+    hdr. ## underlay_id ## _ethernet.setValid(); \
+    hdr. ## underlay_id ## _ethernet.dst_addr = underlay_dmac; \
+    hdr. ## underlay_id ## _ethernet.src_addr = underlay_smac; \
+    hdr. ## underlay_id ## _ethernet.ether_type = IPV4_ETHTYPE; \
+    \
+    hdr. ## underlay_id ## _ipv4.setValid(); \
+    bit<16>  ## overlay_id ## _ip_len = 0; \
+    if (hdr. ## overlay_id ## _ipv4.isValid()) { \
+        ## overlay_id ## _ip_len = ## overlay_id ## _ip_len + hdr. ## overlay_id ## _ipv4.total_len; \
+    } \
+    if (hdr. ## overlay_id ## _ipv6.isValid()) { \
+        ## overlay_id ## _ip_len = (## overlay_id ## _ip_len + IPV6_HDR_SIZE + \
+            hdr. ## overlay_id ## _ipv6.payload_length); \
+    } \
+    hdr. ## underlay_id ## _ipv4.total_len = (ETHER_HDR_SIZE + IPV4_HDR_SIZE + UDP_HDR_SIZE + \
+        NVGRE_HDR_SIZE + ## overlay_id ## _ip_len); \
+    hdr. ## underlay_id ## _ipv4.total_len = (ETHER_HDR_SIZE + IPV4_HDR_SIZE + \
+            NVGRE_HDR_SIZE + hdr. ## underlay_id ## _ipv4.total_len); \
+    hdr. ## underlay_id ## _ipv4.version = 4; \
+    hdr. ## underlay_id ## _ipv4.ihl = 5; \
+    hdr. ## underlay_id ## _ipv4.diffserv = 0; \
+    hdr. ## underlay_id ## _ipv4.identification = 1; \
+    hdr. ## underlay_id ## _ipv4.flags = 0; \
+    hdr. ## underlay_id ## _ipv4.frag_offset = 0; \
+    hdr. ## underlay_id ## _ipv4.ttl = 64; \
+    hdr. ## underlay_id ## _ipv4.protocol = NVGRE_PROTO; \
+    hdr. ## underlay_id ## _ipv4.dst_addr = underlay_dip; \
+    hdr. ## underlay_id ## _ipv4.src_addr = underlay_sip; \
+    hdr. ## underlay_id ## _ipv4.hdr_checksum = 0; \
+    \
+    hdr. ## underlay_id ## _nvgre.setValid(); \
+    hdr. ## underlay_id ## _nvgre.flags = 4; \
+    hdr. ## underlay_id ## _nvgre.reserved = 0; \
+    hdr. ## underlay_id ## _nvgre.version = 0; \
+    hdr. ## underlay_id ## _nvgre.protocol_type = 0x6558; \
+    hdr. ## underlay_id ## _nvgre.vsid = tunnel_key; \
+    hdr. ## underlay_id ## _nvgre.flow_id = 0; \
+}
+#endif
+
+PUSH_VXLAN_TUNNEL_DEF(u0, customer)
+PUSH_VXLAN_TUNNEL_DEF(u1, u0)
+PUSH_NVGRE_TUNNEL_DEF(u0, customer)
+PUSH_NVGRE_TUNNEL_DEF(u1, u0)
+
+#define tunnel_encap(hdr, \
+                    meta, \
+                    overlay_dmac, \
+                    underlay_dmac, \
+                    underlay_smac, \
+                    underlay_dip, \
+                    underlay_sip, \
+                    dash_encapsulation, \
+                    tunnel_key) { \
+    if (dash_encapsulation == dash_encapsulation_t.VXLAN) { \
+        if (meta.tunnel_pointer == 0) { \
+            push_vxlan_tunnel_u0(hdr, \
+                           overlay_dmac, \
+                           underlay_dmac, \
+                           underlay_smac, \
+                           underlay_dip, \
+                           underlay_sip, \
+                           tunnel_key); \
+        } else if (meta.tunnel_pointer == 0) { \
```

Comment:
Typo? Perhaps, it needs to be meta.tunnel_pointer == 1

Reply:
Fixed